pkg/docker/docker.go

```diff
@@ -608,70 +605,81 @@ func (d *stiDocker) RunContainer(opts RunContainerOptions) error {
-		return err
-	}
-
-	// Start the container.
-	glog.V(2).Infof("Starting container %q ...", containerName)
-	if err := util.TimeoutAfter(DefaultDockerTimeout, "timeout after waiting %v for Docker to start container", func() error {
-		return d.client.StartContainer(container.ID, nil)
-	}); err != nil {
-		return err
+		removalDone <- struct{}{}
 	}
+	dumpStack := func(signal os.Signal) {
+		if signal == syscall.SIGQUIT {
+			buf := make([]byte, 1<<16)
+			runtime.Stack(buf, true)
+			fmt.Printf("%s", buf)
+		}
+		os.Exit(1)
+	}
+	return util.NewInterruptHandler(dumpStack, removeContainer).Run(func() error {
+		// Attach to the container.
+		glog.V(2).Infof("Attaching to container %q ...", containerName)
+		attachOpts := opts.asDockerAttachToContainerOptions()
+		attachOpts.Container = container.ID
+		if _, err = d.client.AttachToContainerNonBlocking(attachOpts); err != nil {
+			glog.Errorf("Unable to attach to container %q with options %+v: %v", containerName, attachOpts, err)
+			return err
+		}
 
-	// Run OnStart hook if defined.
-	if opts.OnStart != nil {
-		if err = opts.OnStart(container.ID); err != nil {
+		// Start the container.
+		glog.V(2).Infof("Starting container %q ...", containerName)
+		if err := util.TimeoutAfter(DefaultDockerTimeout, "timeout after waiting %v for Docker to start container", func() error {
+			return d.client.StartContainer(container.ID, nil)
+		}); err != nil {
 			return err
 		}
-	}
 
-	// We either block waiting for a user-originated SIGINT, or wait for the
-	// container to terminate. When TargetImage is true, we're dealing with
-	// an invocation of `s2i build ... --run` so this will, e.g., run a web
-	// server and block until the user interrupts it. The other case is seen
-	// when running the assemble script or other commands that are meant to
-	// terminate in a finite amount of time.
-	if opts.TargetImage {
-		runContainerDockerRun(container, d, image)
-	} else {
-		// Return an error if the exit code of the container is
-		// non-zero.
-		glog.V(4).Infof("Waiting for container %q to stop ...", containerName)
-		exitCode, err := d.client.WaitContainer(container.ID)
-		if err != nil {
-			return fmt.Errorf("waiting for container %q to stop: %v", containerName, err)
+		// Run OnStart hook if defined.
+		if opts.OnStart != nil {
+			if err = opts.OnStart(container.ID); err != nil {
+				return err
+			}
 		}
-		if exitCode != 0 {
-			return errors.NewContainerError(container.Name, exitCode, "")
+
+		// We either block waiting for a user-originated SIGINT, or wait for the
+		// container to terminate. When TargetImage is true, we're dealing with
+		// an invocation of `s2i build ... --run` so this will, e.g., run a web
+		// server and block until the user interrupts it. The other case is seen
+		// when running the assemble script or other commands that are meant to
+		// terminate in a finite amount of time.
+		if opts.TargetImage {
+			runContainerDockerRun(container, d, image)
```

Comment:
Wait, "wait forever" and "block forever" is not the same thing, is it?

If it blocks forever, then this line will block forever as well?